src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -0,0 +1,257 @@
+/*-------------------------------------------------------------------------
+ *
+ * intermediate_result_pruning.c
+ *   Functions for pruning intermediate result broadcasting.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "distributed/citus_custom_scan.h"
+#include "distributed/intermediate_result_pruning.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/worker_manager.h"
+#include "utils/builtins.h"
+
+static List * AppendAllAccessedWorkerNodes(List *workerNodeList,
+										   DistributedPlan *distributedPlan,
+										   int workerNodeCount);
+static IntermediateResultsHashEntry * SearchIntermediateResult(HTAB
+															   *intermediateResultsHash,
+															   char *resultId);
+
+
+/*
+ * FindSubPlansUsedInPlan adds all the subplans used by the plan into
+ * plan->usedSubPlanNodeList field. Simply traversing the range table entries in the plan.
+ */
+void
+FindSubPlansUsedInPlan(DistributedPlan *plan)
+{
+	Query *jobQuery = NULL;
+	List *rangeTableList = NIL;
+	ListCell *rangeTableCell = NULL;
+
+	jobQuery = plan->workerJob == NULL ? plan->insertSelectSubquery :
+			   plan->workerJob->jobQuery;
+
+	/*
+	 * Decide if there are no intermediate results.
+	 *
+	 * We usually store the intermediate result info in workerJob field. We currently
+	 * have a single exception to this, which is SELECT INTO queries that have
+	 * ON CONFLICT/RETURNING clauses. If the query is of such a query we need to consider
+	 * pruning intermediate results even though we do not have a workerjob field
+	 *
+	 *  * We do not wish to proceed if one of the following is met:
+	 * - there are no intermediate results
+	 * - the query hits all the worker nodes
+	 */
+	if (jobQuery == NULL)
+	{
+		return;
+	}
+
+	rangeTableList = ExtractRangeTableEntryList(jobQuery);
+
+	foreach(rangeTableCell, rangeTableList)
+	{
+		RangeTblEntry *rangeTableEntry = lfirst(rangeTableCell);
+		if (rangeTableEntry->rtekind == RTE_FUNCTION)
+		{
+			Const *resultIdConst = NULL;
+			Datum resultIdDatum = 0;
+			char *resultId = NULL;
+
+			resultIdConst = FindResultIdOfIntermediateResultFunction(rangeTableEntry);
+
+			if (resultIdConst == NULL)
+			{
+				continue;
+			}
+
+			plan->usedSubPlanNodeList = list_append_unique(plan->usedSubPlanNodeList,
+														   resultIdConst);
+
+			resultIdDatum = resultIdConst->constvalue;
+			resultId = TextDatumGetCString(resultIdDatum);
+			elog(DEBUG4, "appending %s subplan to usedSubPlanNodeList of plan %lu",
+				 resultId, plan->planId);
```

Comment:
minor: it's a little confusing that we say "appending...", but we actually already appended. Maybe move the append below?